File: 74.cpp
/**
 * class VersionControl {
 *     public:
 *     static bool isBadVersion(int k);
 * }
 * you can use VersionControl::isBadVersion(k) to judge wether
 * the kth code version is bad or not.
*/
class Solution {
public:
    /**
     * @param n: An integers.
     * @return: An integer which is the first bad version.
     */
    int findFirstBadVersion(int n) {
        if (n < 1) {
            return -1;
        }

        
        int mid,left=1,right=n;
    
        while(left<=right)
        {
            mid=left+(right-left)/2;//这步很重要 不要溢出
            cout<<mid<<endl;
            if(!VersionControl::isBadVersion(mid))
                right=mid-1;
            else
                left=mid+1;
        }


        return left;   

        return -1;  // find no bad version
    }
};

File: 74.py
#class SVNRepo:
#    @classmethod
#    def isBadVersion(cls, id)
#        # Run unit tests to check whether verison `id` is a bad version
#        # return true if unit tests passed else false.
# You can use SVNRepo.isBadVersion(10) to check whether version 10 is a 
# bad version.
class Solution:
    """
    @param n: An integers.
    @return: An integer which is the first bad version.
    """
    def findFirstBadVersion(self, n):
        # write your code here
        l=1
        r=n
        mid=1;
        if (n==1):
            return 1;
        while (l<=r):
            mid=(l+r)/2;
            if (SVNRepo.isBadVersion(mid)):
                r=mid-1;
            else:
                l=mid+1;

        if (mid>1 and SVNRepo.isBadVersion(mid-1)==True):
            return mid-1;
        if (mid>=1 and SVNRepo.isBadVersion(mid)==True):
            return mid;
        if (mid<n and SVNRepo.isBadVersion(mid+1)==True):
            return mid+1;

        # Total Runtime: 300 ms
